mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java

```diff
@@ -90,50 +91,65 @@ public boolean isHandling(Event event) {
+    public Publisher<Void> reactiveEvent(Event event) {
         Username username = Username.of(getClass().getCanonicalName());
         if (event instanceof MessageMoveEvent) {
             MailboxSession session = mailboxManager.createSystemSession(username);
-            handleMessageMove(event, session, (MessageMoveEvent) event);
+            return handleMessageMove(event, session, (MessageMoveEvent) event);
         }
         if (event instanceof Added) {
             MailboxSession session = mailboxManager.createSystemSession(username);
-            handleAdded(event, session, (Added) event);
+            return handleAdded(event, session, (Added) event);
         }
+        return Mono.empty();
     }
 
-    private void handleAdded(Event event, MailboxSession session, Added addedEvent) throws MailboxException {
+    private Mono<Void> handleAdded(Event event, MailboxSession session, Added addedEvent) {
         if (isAppendedToInbox(addedEvent)) {
-            Mailbox mailbox = mapperFactory.getMailboxMapper(session).findMailboxById(addedEvent.getMailboxId());
             MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
 
-            List<InputStream> contents = MessageRange.toRanges(addedEvent.getUids())
-                .stream()
-                .flatMap(range -> retrieveMessages(messageMapper, mailbox, range))
-                .map(Throwing.function(MailboxMessage::getFullContent))
-                .collect(Guavate.toImmutableList());
-            spamAssassin.learnHam(contents, event.getUsername());
+           return mapperFactory.getMailboxMapper(session)
+                .findMailboxByIdReactive(addedEvent.getMailboxId())
+                .flatMapMany(mailbox -> retrieveContents(addedEvent, mailbox, messageMapper))
+                .flatMap(contents -> spamAssassin.learnHam(contents, event.getUsername()))
+               .then();
         }
+        return Mono.empty();
     }
 
-    private void handleMessageMove(Event event, MailboxSession session, MessageMoveEvent messageMoveEvent) {
-        if (isMessageMovedToSpamMailbox(messageMoveEvent)) {
-            LOGGER.debug("Spam event detected");
-            ImmutableList<InputStream> messages = retrieveMessages(messageMoveEvent, session);
-            spamAssassin.learnSpam(messages, event.getUsername());
-        }
-        if (isMessageMovedOutOfSpamMailbox(messageMoveEvent)) {
-            ImmutableList<InputStream> messages = retrieveMessages(messageMoveEvent, session);
-            spamAssassin.learnHam(messages, event.getUsername());
-        }
+    private Flux<MessageToLearn> retrieveContents(Added addedEvent, Mailbox mailbox, MessageMapper messageMapper) {
+        return Flux.defer(() -> Flux.fromIterable(MessageRange.toRanges(addedEvent.getUids())))
+                .flatMap(range -> retrieveMessages(messageMapper, mailbox, range))
+                .map((ThrowingFunction<MailboxMessage, MessageToLearn>) message ->
```

Comment:
why do you need to wrap into a `ThrowingFunction`?

Reply:
because `message.getFullContent()` can throw an IOException

Reply:
if it can throw IOException it's doing IO and if it's doing IO you should wrap that into a Mono + flatMap, right?